### DISCUSSION TITLE
get path to gdl executable

### DISCUSSION THREAD

**Discussion by alaingdl** [gnudatalanguage/gdl#1719] [General]
OK, the test 'test_shmmap' is working fine on Linux and OSX if I can provide the path to the GDL executable, and I know how to do for Linux & OSX in C in few lines (tested) and should be easy to extend it to others OS.

Could you suggested me a name for such non-standard function ? GET_PATH_GDL_EXE() ?

(a generalization of the Linux-only code : 
get_pid=FILE_READLINK('/proc/self')
get_abspath_to_exe=file_readlink('/proc/'+get_pid+'/exe')
)

One drawback is that now the SHMMAP test is working fine in IDL & GDL but we can circumvent that quite easily for Linux ...

another solution is to just do the C code only for OSX ? (because I don't know how to do it just by call to GDL syntax code ...)

**Comment by alaingdl** [comment]
OK, as long at it is not needed in the code, I just improved get_path_to_exe() in testsuite/ which can detect trivially GDL, FL & IDL on Linux, and IDL & GDL on OSX. Then some tests previously using ../src/gdl can run anywhere now. And I can test with success the code `test_shmmap` for GDL, FL & IDL on Linux, and  for IDL & GDL on OSX (as mentioned in #1715 types are not OK in GDL and yes, the test should be extended, but the proof of concept is OK :smile:

**Comment by GillesDuvert** [comment]
@alaingdl test_shmap seems to have problems on OSX (last 2 Prs at least)

**Reply by alaingdl** [reply]
And I cannot anymore compile plplot using the script on OSX :(

```
$ ./scripts/build_gdl.sh prep
[...]
CMake Error at cmake/modules/plplot_functions.cmake:619 (set_target_properties):
  set_target_properties called with incorrect number of arguments.
Call Stack (most recent call first):
  src/CMakeLists.txt:301 (configure_library_build)
```

**Reply by alaingdl** [reply]
@GillesDuvert 
After some tests, yes the code in `shm.cpp` and in `shm.hpp` was changed between Feb 13 & 15 and now SHMMAP and the related test are not working anymore on my U2204 and my Debian 10

HTH

```
diff -b gdl-1.0.4git240213CMake/src/shm.cpp gdl-1.0.4git240215CMake/src/shm.cpp
247c247
< 	  if ((*i).second.flags & USE_SYSV == USE_SYSV) {
---
> 	  if (((*i).second.flags & USE_SYSV) == USE_SYSV) {
399c399
< 		  if ((*i).second.flags & USE_SYSV == USE_SYSV) {
---
> 		  if (((*i).second.flags & USE_SYSV) == USE_SYSV) {

$ diff gdl-1.0.4git240213CMake/src/shm.hpp gdl-1.0.4git240215CMake/src/shm.hpp
27c27
<   void* mapped_address=NULL;
---
>   void* mapped_address;
29,31c29,31
<   size_t length=0;
<   off_t offset=0;
<   int   refcount=0;
---
>   size_t length;
>   off_t offset;
>   int   refcount;
33,34c33,34
<   int type=0;
<   int flags=0;
---
>   int type;
>   int flags;
```

**Reply by GillesDuvert** [reply]
well, this is it. Amusing, as this change was requires by the OSX compiler.

**Reply by GillesDuvert** [reply]
OK, an element was not initialized in the .cpp, as the OSX compiler did not appreciate it to be initialized in the .hpp . Cpluplus subtetly...
